ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java

```diff
@@ -0,0 +1,228 @@
+    NavigableMap<String, Long> validSizePerLogSegment =
+        collectValidDataSizeByLogSegment(deleteReferenceTimeInMs, expirationReferenceTimeInMs);
+    return new Pair<>(deleteReferenceTimeInMs, validSizePerLogSegment);
+  }
+
+  /**
+   * Gets the size of valid data for all serviceIds and their containerIds as of now (the time when the API is called).
+   * The following data are considered as valid data for this API:
+   * 1. PUT with no expiry and no corresponding DELETE
+   * 2. PUT expiring at t_exp but t_ref < t_exp
+   * 3. PUT with DELETE at time t_delete but t_ref < t_delete
+   * @return the valid data size for each container in the form of a nested {@link Map} of serviceIds to another map of
+   * containerIds to valid data size.
+   */
+  Map<String, Map<String, Long>> getValidDataSizeByContainer() throws StoreException {
+    long deleteAndExpirationRefTimeInMs = time.milliseconds();
+    return collectValidDataSizeByContainer(deleteAndExpirationRefTimeInMs);
+  }
+
+  /**
+   * Walk through the entire index and collect valid data size information per container (size of delete records not included).
+   * @param deleteAndExpirationRefTimeInMs the reference time in ms until which deletes and expiration are relevant
+   * @return a nested {@link Map} of serviceId to containerId to valid data size
+   */
+  private Map<String, Map<String, Long>> collectValidDataSizeByContainer(long deleteAndExpirationRefTimeInMs)
+      throws StoreException {
+    Set<String> deletedKeys = new HashSet<>();
+    Map<String, Map<String, Long>> validDataSizePerContainer = new HashMap<>();
+    for (IndexSegment indexSegment : index.getIndexSegments().descendingMap().values()) {
+      List<IndexValue> validIndexValues =
+          getValidIndexValues(indexSegment, deleteAndExpirationRefTimeInMs, deleteAndExpirationRefTimeInMs,
+              deletedKeys);
+      for (IndexValue value : validIndexValues) {
+        if (!value.isFlagSet(IndexValue.Flags.Delete_Index)) {
+          // delete records does not count towards valid data size for quota (containers)
+          updateNestedMapHelper(validDataSizePerContainer, String.valueOf(value.getServiceId()),
+              String.valueOf(value.getContainerId()), value.getSize());
+        }
+      }
+    }
+    return validDataSizePerContainer;
+  }
+
+  /**
+   * Walk through the entire index and collect valid size information per log segment (size of delete records included).
+   * @param deleteReferenceTimeInMs the reference time in ms until which deletes are relevant
+   * @param expirationReferenceTimeInMs the reference time in ms until which expiration are relevant
+   * @return a {@link NavigableMap} of log segment name to valid data size
+   */
+  private NavigableMap<String, Long> collectValidDataSizeByLogSegment(long deleteReferenceTimeInMs,
+      long expirationReferenceTimeInMs) throws StoreException {
+    Set<String> deletedKeys = new HashSet<>();
```

Comment:
Originally, I thought we could extract just the uuid as a String. Since that's not available, I will change it to StoreKey.